Parametric Heteroscedastic Example

willtebbutt opened the pull request:
<!-- Comment lines like this one will remain invisible -->

<!-- Thank you for your contribution! Please fill in this template so that we
can understand your intent and the proposed changes. If anything about this
template is unclear, just mention it! -->

**Summary**
<!-- Summary of what & why - explain your motivation and/or link to any GitHub issues this relates to -->
An example that shows off one of the benefits of allowing the user to specify the observation variance.
This is only marginally more involved than a standard regression example.
This model came up when I was chatting to @tpgillam , and it seemed sufficiently simple to implement that I might as well just do it.

**Proposed changes**
<!-- Large PRs should ideally be preceded by a design discussion on a separate issue! -->

<!-- A clear and concise description of the contents of this pull request. -->
N/A


**What alternatives have you considered?**
<!-- A clear and concise description of any alternative solutions or features you've considered. -->
N/A

**Breaking changes**
<!-- If this PR breaks backwards-compatibility, please start the PR title with `**BREAKING**`! -->
<!-- In this section, describe any changes that are not backwards-compatible, -->
<!-- why it is worth breaking backwards compatiblity, -->
<!-- and how a user would have to address these changes in their downstream code. -->
N/A

theogf commented on examples/parametric-heteroscedastic/script.jl:
```diff
@@ -0,0 +1,82 @@
+
+# Build inputs.
+const x = range(0.0, 10.0; length=100)
+const y = rand(Xoshiro(123456), build_gp(θ_init)(x, observation_variance(θ_init, x)))
```
```suggestion
const y = rand(Xoshiro(123456), build_gp(θ_init)(x, observation_variance(θ_init, x)));
```

willtebbutt replied:
Have resolved locally.

theogf commented on examples/parametric-heteroscedastic/script.jl:
```diff
@@ -0,0 +1,82 @@
+    f = build_gp(θ)
+    Σ = observation_variance(θ, x)
+    return -logpdf(f(x, Σ), y)
+end
```
```suggestion
end;
```

willtebbutt replied:
Have resolved locally.

theogf commented on examples/parametric-heteroscedastic/script.jl:
```diff
@@ -0,0 +1,82 @@
+    return -logpdf(f(x, Σ), y)
+end
+
+# Optimise the hyperparameters. They've been initialised to the correct values, so
```
Maybe it's worth to shift them by 1 in log-space or something like this

theogf commented on examples/parametric-heteroscedastic/script.jl:
```diff
@@ -0,0 +1,82 @@
+let
+    fig = Figure()
+    ax = Axis(fig[1, 1])
+    plot!(ax, x, f_post(x, Σ_obs_final); bandscale=3)
```
Maybe have some labels so we know what we are looking at on the plot

willtebbutt replied:
Yup, that would be helpful. Am doing.

willtebbutt commented:
Hmm @theogf for some reason the docs build isn't updating properly, so I can't look at the changes directly. I'm confident that I've addressed your concerns though -- are you happy to approve, or would you prefer to figure out why we're not getting the right preview?

theogf commented:
@willtebbutt Now there is always a delay for gh-pages because of this action: https://github.com/JuliaGaussianProcesses/AbstractGPs.jl/actions/workflows/pages/pages-build-deployment

willtebbutt commented:
@theogf seems to be doing something right now.

theogf commented:
Sorry  to be picky but could you put a much higher alpha (0.4 or 0.5) for the posterior + noise, I could not see anything at first 
![image](https://user-images.githubusercontent.com/11480067/158433146-5b5c569d-f223-46e9-a0b5-b3eecd146ebe.png)

willtebbutt commented:
Np. Have upped it. Will merge when CI passes, subject to the plot not needing another tweak.
